src/main/java/com/swvl/geometry/shapes/Polygon.java

```diff
@@ -0,0 +1,408 @@
+
+        /* Calculate vector qr^ */
+        double qrx = r.x - q.x;
+        double qry = r.y - q.y;
+
+        /* Cross product between pq^ and qr^ */
+        double val = pqx * qry - pqy * qrx;
+
+        /* if val = 0 then points are collinear */
+        return val < Point.EPS && val > -Point.EPS;
+
+    }
+
+    @Override
+    public Polygon clone() {
+        Point[] clonedPoints = new Point[points.length];
+        for (int i = 0; i < points.length - 1; ++i)
+            clonedPoints[i] = points[i].clone();
+
+        return new Polygon(clonedPoints);
+    }
+
+    /**
+     * Calculates the centroid of the polygon's vertices (not guaranteed that centroid exist inside the polygon)
+     */
+    @Override
+    public Point getCenterPoint() {
+        double xSum = 0, ySum = 0;
+        for (Point point : points) {
+            xSum += point.x;
+            ySum += point.y;
+        }
+
+        return new Point(xSum / (points.length - 1),
+                ySum / (points.length - 1));
+    }
+
+    @Override
+    public boolean contains(Shape shape) throws OperationNotSupportedException {
+        if (!isInitialized)
+            init(points);
+
+        if (shape instanceof Point)
+            return isPointIntersection((Point) shape);
+
+        if (shape instanceof LineSegment) {
+            LineSegment line = (LineSegment) shape;
+            return isPointIntersection(line.getStartPoint())
+                    && isPointIntersection(line.getEndPoint());
+        }
+
+        if (shape instanceof Rectangle)
+            return containsRectangle((Rectangle) shape);
+
+        if (shape instanceof Polygon)
+            return containsPolygon((Polygon) shape);
+
+        throw new OperationNotSupportedException("contains operation in Polygon does not support " + shape.getClass());
+
+    }
+
+    private boolean containsRectangle(Rectangle rect) throws OperationNotSupportedException {
+        Point p1 = new Point(rect.maxPoint.x, rect.minPoint.y); // bottom-right point
+        Point p2 = new Point(rect.minPoint.x, rect.maxPoint.y); // upper-left point
+
+        /* Edges of a rectangle */
+        Point[] rectPoints = new Point[]{
+                rect.minPoint,
+                p1,
+                rect.maxPoint,
+                p2
+        };
+
+        boolean isContained = true;
+
+        /*
+         * Iterate over rectangle points and check if that all points are
+         * inside the polygon
+         */
+        for (Point rectPoint : rectPoints)
+            isContained &= this.isPointIntersection(rectPoint);
+
+        return isContained;
+    }
+
+    private boolean containsPolygon(Polygon polygon) throws OperationNotSupportedException {
+        boolean isContained = true;
+
+        /* Iterate over polygon's points and check if any point is inside the invoker polygon*/
+        for (int i = 0; i < polygon.points.length - 1; ++i)
+            isContained &= this.isPointIntersection(polygon.points[i]);
```

Comment:
save time and return if false. I suggest naming the subcall `containsPoint` for confusion